File: Coding_Problems_Java/Keys and rooms.java
class Solution {
    private void dfs(int node, int[] visited, List<List<Integer>> rooms){
        visited[node] = 1;
        List<Integer> adj_room = rooms.get(node);

        for(int room: adj_room){
            if(visited[room]==0){
                dfs(room, visited, rooms);
            }
        }
    }
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        int n = rooms.size();
        int [] visited = new int[n];

        dfs(0, visited, rooms);
        for(int i = 0; i < n; i++){
            if (visited[i] == 0){
                return false;
            }
        }
        return true;
    }
    
}

File: Coding Problems - Python/Graphs/Keys and Rooms.py
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        def dfs(node):
            visited[node] = 1
            adj_room = rooms[node]

            for room in adj_room:
                if not visited[room]:
                    dfs(room)

        n = len(rooms)
        visited = [0]*n

        dfs(0)
        
        for i in range(n):
            if not visited[i]:
                return False
        return True
            
